### ISSUE TITLE
Resume training performance drop

### ISSUE THREAD

**Issue by zhixuanli** [hszhao/semseg#6]
Hi, I'm trying to use your fantastic code but encountered something confused when I try to load a checkpoint of 200 epoch training.

The mIoU of training reached 0.739 at epoch 200. But when I try to load this checkpoint and continually training, the performance dropped from 0.739 to 0.563 at epoch 201.

It's quite hard for me to understand this phenomenon, could you give me some advices? 
Thank you so much!

## Here is my config file:
```
DATA:
  data_root: /datasets-ssd/cityscapes
  train_list: dataset/cityscapes/list/fine_train.txt
  val_list: dataset/cityscapes/list/fine_val.txt
  classes: 19

TRAIN:
  arch: psp
  layers: 101
  sync_bn: True  # adopt syncbn or not
  train_h: 513
  train_w: 513
  scale_min: 0.5  # minimum random scale
  scale_max: 2.0  # maximum random scale
  rotate_min: -10  # minimum random rotate
  rotate_max: 10  # maximum random rotate
  zoom_factor: 8  # zoom factor for final prediction during training, be in [1, 2, 4, 8]
  ignore_label: 255
  aux_weight: 0.4
#  train_gpu: [0, 1, 2, 3, 4, 5, 6, 7]
  train_gpu: [0]
  workers: 6  # data loader workers
  batch_size: 2  # batch size for training
  batch_size_val: 1  # batch size for validation during training, memory and speed tradeoff
  base_lr: 0.01
  epochs: 400
  start_epoch: 200
  power: 0.9
  momentum: 0.9
  weight_decay: 0.0001
  manual_seed: 520
  print_freq: 10
  save_freq: 1
  save_path: exp/cityscapes/pspnet101/model
  weight:  # path to initial weight (default: none)
  resume:  exp/cityscapes/pspnet101/model/train_epoch_200.pth # path to latest checkpoint (default: none)
  evaluate: False  # evaluate on validation set, extra gpu memory needed and small batch_size_val is recommend
Distributed:
  distributed: True
  dist_url: tcp://127.0.0.1:6789
  dist_backend: 'nccl'
  multiprocessing_distributed: True
  world_size: 1
  rank: 0
  use_apex: True
  opt_level: 'O0'
  keep_batchnorm_fp32:
  loss_scale:

TEST:
  test_list: dataset/cityscapes/list/fine_val.txt
  split: val  # split in [train, val and test]
  base_size: 2048  # based size for scaling
  test_h: 713
  test_w: 713
  scales: [1.0]  # evaluation scales, ms as [0.5, 0.75, 1.0, 1.25, 1.5, 1.75]
  has_prediction: False  # has prediction already or not
  index_start: 0  # evaluation start index in list
  index_step: 0  # evaluation step index in list, 0 means to end
  test_gpu: [0]
  model_path: exp/cityscapes/pspnet101/model/train_epoch_400.pth  # evaluation model path
  save_folder: exp/cityscapes/pspnet101/result/epoch_400/val/ss  # results save folder
  colors_path: dataset/cityscapes/cityscapes_colors.txt  # path of dataset colors
  names_path: dataset/cityscapes/cityscapes_names.txt  # path of dataset category names

```

## My environment:
+ 2080ti, 12G
+ CUDA 10.0
+ python 3.6

Hoping for your reply!
Thank you again!

**Comment by hszhao** [comment]
Hi, is you 200 epoch model `train_epoch_200.pth' trained with epochs set to 200 at the beginning? If yes, you cannot resume from this checkpoint for longer training epochs. The learning rate will change a lot if you just set the epochs to 400 in the yaml file. In this case, you need to train from the beginning.

**Comment by zhixuanli** [comment]
Thank you!

Another question, can your code running cross server nodes? 
Our school has a common-server for us to use, but each node only has 4 gpus. And as you have instructed, 8 gpus should be the best configuration.

Any ideas about this? 

And I have sent you an email many days ago, asking for how to write cuda programs. Can you recommend some books or tutorials? 
Thank you again for your open-source code and your patience.

**Comment by zhixuanli** [comment]
```
Distributed:
  distributed: True
  dist_url: tcp://127.0.0.1:6789
  dist_backend: 'nccl'
  multiprocessing_distributed: True
  world_size: 1
  rank: 0
  use_apex: True
  opt_level: 'O0'
  keep_batchnorm_fp32:
  loss_scale:
```

What's the meaning of the following two parameters?
```
  world_size: 1
  rank: 0
```

Thanks!

**Comment by hszhao** [comment]
1. I have not tested using multiple nodes yet, you can have a look at the apex usage or official pytorch usage (then you need to use the official sync bn).
2. For the cuda code, I wrote lots of them during my undergraduate study. I think it is the C style code and easy to learn, you need to know the design logic of GPU threads.

**Comment by zhixuanli** [comment]
Thanks for your reply.
Can you tell the meaning of 
```
  world_size: 1
  rank: 0
  opt_level: 'O0'
```
in the config file?

**Comment by zhixuanli** [comment]
@hszhao 
Sorry for bothering you again. Could you please give me an answer about the above question?

Thank so much!
